modules/Evse15118D20/charger/ISO15118_chargerImpl.cpp

```diff
@@ -0,0 +1,387 @@
+
+#include <iso15118/io/logging.hpp>
+#include <iso15118/session/logger.hpp>
+
+namespace module {
+namespace charger {
+
+static constexpr auto WAIT_FOR_SETUP_DONE_MS{std::chrono::milliseconds(200)};
+
+std::mutex GEL; // Global EVerest Lock
+
+namespace {
+
+std::filesystem::path construct_cert_path(const std::filesystem::path& initial_path, const std::string& config_path) {
+    if (config_path.empty()) {
+        return initial_path;
+    }
+
+    if (config_path.front() == '/') {
+        return config_path;
+    } else {
+        return initial_path / config_path;
+    }
+}
+
+iso15118::config::TlsNegotiationStrategy convert_tls_negotiation_strategy(const std::string& strategy) {
+    using Strategy = iso15118::config::TlsNegotiationStrategy;
+    if (strategy == "ACCEPT_CLIENT_OFFER") {
+        return Strategy::ACCEPT_CLIENT_OFFER;
+    } else if (strategy == "ENFORCE_TLS") {
+        return Strategy::ENFORCE_TLS;
+    } else if (strategy == "ENFORCE_NO_TLS") {
+        return Strategy::ENFORCE_NO_TLS;
+    } else {
+        EVLOG_AND_THROW(Everest::EverestConfigError("Invalid choice for tls_negotiation_strategy: " + strategy));
+        // better safe than sorry
+    }
+}
+
+types::iso15118_charger::DisplayParameters
+convert_display_parameters(const iso15118::session::feedback::DisplayParameters& in) {
+    return {in.present_soc,
+            in.minimum_soc,
+            in.target_soc,
+            in.maximum_soc,
+            in.remaining_time_to_minimum_soc,
+            in.remaining_time_to_target_soc,
+            in.remaining_time_to_maximum_soc,
+            in.battery_energy_capacity,
+            in.inlet_hot};
+}
+
+} // namespace
+
+void ISO15118_chargerImpl::init() {
+
+    // setup logging routine
+    iso15118::io::set_logging_callback([](const iso15118::LogLevel& level, const std::string& msg) {
+        switch (level) {
+        case iso15118::LogLevel::Error:
+            EVLOG_error << msg;
+            break;
+        case iso15118::LogLevel::Warning:
+            EVLOG_warning << msg;
+            break;
+        case iso15118::LogLevel::Info:
+            EVLOG_info << msg;
+            break;
+        case iso15118::LogLevel::Debug:
+            EVLOG_debug << msg;
+            break;
+        case iso15118::LogLevel::Trace:
+            EVLOG_verbose << msg;
+            break;
+        default:
+            EVLOG_critical << "(Loglevel not defined) - " << msg;
+            break;
+        }
+    });
+}
+
+void ISO15118_chargerImpl::ready() {
+
+    while (true) {
+        if (setup_steps_done.all()) {
+            break;
+        }
+        std::this_thread::sleep_for(WAIT_FOR_SETUP_DONE_MS);
+    }
+
+    const auto session_logger = std::make_unique<SessionLogger>(mod->config.logging_path);
+
+    const auto default_cert_path = mod->info.paths.etc / "certs";
+    const auto cert_path = construct_cert_path(default_cert_path, mod->config.certificate_path);
+    const iso15118::TbdConfig tbd_config = {
```

Comment:
What does Tbd stand for?

Reply:
Tbd -> To be discussed. I will change that in libiso15118 in the near future :+1: